SmartDeviceLink/SDLChoiceSetManager.m

```diff
@@ -459,18 +513,79 @@ - (SDLKeyboardProperties *)sdl_defaultKeyboardConfiguration {
 
 #pragma mark - Getters
 
+- (NSMutableSet<SDLChoiceCell *> *)preloadedMutableChoices {
+    __block NSMutableSet<SDLChoiceCell *> *set = nil;
+    [self sdl_runSyncOnQueue:^{
+        set = self->_preloadedMutableChoices;
+    }];
+
+    return set;
+}
+
+- (NSMutableSet<SDLChoiceCell *> *)pendingMutablePreloadChoices {
+    __block NSMutableSet<SDLChoiceCell *> *set = nil;
+    [self sdl_runSyncOnQueue:^{
+        set = self->_pendingMutablePreloadChoices;
+    }];
+
+    return set;
+}
+
 - (NSSet<SDLChoiceCell *> *)preloadedChoices {
-    return [_preloadedMutableChoices copy];
+    __block NSSet<SDLChoiceCell *> *set = nil;
+    [self sdl_runSyncOnQueue:^{
+        set = [self->_preloadedMutableChoices copy];
+    }];
+
+    return set;
 }
 
 - (NSSet<SDLChoiceCell *> *)pendingPreloadChoices {
-    return [_pendingMutablePreloadChoices copy];
+    __block NSSet<SDLChoiceCell *> *set = nil;
+    [self sdl_runSyncOnQueue:^{
+        set = [self->_pendingMutablePreloadChoices copy];
+    }];
+
+    return set;
+}
+
+- (UInt16)nextChoiceId {
+    __block UInt16 choiceId = 0;
+    [self sdl_runSyncOnQueue:^{
+        choiceId = self->_nextChoiceId;
+        self->_nextChoiceId = choiceId + 1;
+    }];
+
+    return choiceId;
+}
+
+- (UInt16)nextCancelId {
+    __block UInt16 cancelId = 0;
+    [self sdl_runSyncOnQueue:^{
+        cancelId = self->_nextCancelId;
+        self->_nextCancelId = cancelId + 1;
+    }];
+
+    return cancelId;
 }
 
 - (NSString *)currentState {
     return self.stateMachine.currentState;
 }
 
+#pragma mark Utilities
+
+/// Checks if we are already on a serial queue. If so, the block is added to the queue; if not, the block is dispatched to the serial `readWrite` queue.
+/// @discussion Used to synchronize access to class properties.
+/// @param block The block to be executed.
+- (void)sdl_runSyncOnQueue:(void (^)(void))block {
+    if (dispatch_get_specific(SDLProcessingQueueName) != nil) {
+        block();
+    } else {
+        dispatch_sync(self.readWriteQueue, block);
+    }
+}
```

Comment:
We use this a number of places now, should be put it on `SDLGlobals` instead?

Reply:
There are four other classes that use the same logic. It would be nice to move the logic to `SDLGlobals`. The only downside is that each class gives its "readWriteQueue" a custom label that makes it easier to debug issues. Perhaps it would be better to create a separate issue for this feature.

Reply:
I agree, we can make this a separate issue.

Reply:
Created PR #1614